service/collector/trusted_advisor.go

```diff
@@ -0,0 +1,240 @@
+			return
+		}
+
+		c.logger.Log("level", "error", "message", "could not get Trusted Advisor checks", "stack", fmt.Sprintf("%#v", err))
+		getChecksError.Inc()
+		return
+	}
+	ch <- trustedAdvisorSupported()
+
+	var wg sync.WaitGroup
+
+	for _, check := range checks {
+		// Ignore any checks that don't relate to service limits.
+		if *check.Category != serviceLimitCategory {
+			continue
+		}
+
+		wg.Add(1)
+
+		go func(id *string) {
+			defer wg.Done()
+
+			resources, err := c.getTrustedAdvisorResources(id)
+			if err != nil {
+				c.logger.Log("level", "error", "message", "could not get Trusted Advisor resource", "stack", fmt.Sprintf("%#v", err), "id", *id)
+				getResourcesError.Inc()
+				return
+			}
+
+			for _, resource := range resources {
+				// One Trusted Advisor check returns the nil string for current usage. Skip it.
+				if len(resource.Metadata) == 6 && resource.Metadata[4] == nil {
+					continue
+				}
+
+				limit, usage, err := resourceToMetrics(resource)
+				if err != nil {
+					c.logger.Log("level", "error", "message", "could not convert Trusted Advisor resource into metrics", "stack", fmt.Sprintf("%#v", err), "id", *id)
+					convertResourcesError.Inc()
+					return
+				}
+
+				ch <- limit
+				ch <- usage
+			}
+		}(check.Id)
+	}
+
+	wg.Wait()
+
+	c.logger.Log("level", "debug", "message", "finished collecting metrics for trusted advisor checks")
+}
+
+// getTrustedAdvisorCheckDescriptions calls Trusted Advisor API to get all available checks.
+func (c *Collector) getTrustedAdvisorChecks() ([]*support.TrustedAdvisorCheckDescription, error) {
+	timer := prometheus.NewTimer(getChecksDuration)
+
+	englishLanguage := "en"
+	describeChecksInput := &support.DescribeTrustedAdvisorChecksInput{
+		Language: &englishLanguage,
+	}
+	describeChecksOutput, err := c.awsClients.Support.DescribeTrustedAdvisorChecks(describeChecksInput)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	timer.ObserveDuration()
+
+	return describeChecksOutput.Checks, nil
+}
+
+// getTrustedAdvisorResources calls Trusted Advisor API to get flagged resources of the given check ID.
+func (c *Collector) getTrustedAdvisorResources(id *string) ([]*support.TrustedAdvisorResourceDetail, error) {
+	timer := prometheus.NewTimer(getResourcesDuration)
+
+	checkResultInput := &support.DescribeTrustedAdvisorCheckResultInput{
+		CheckId: id,
+	}
+	checkResultOutput, err := c.awsClients.Support.DescribeTrustedAdvisorCheckResult(checkResultInput)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	timer.ObserveDuration()
+
+	return checkResultOutput.Result.FlaggedResources, nil
+}
+
+func resourceToMetrics(resource *support.TrustedAdvisorResourceDetail) (prometheus.Metric, prometheus.Metric, error) {
+	region := resource.Metadata[regionIndex]
+	service := resource.Metadata[serviceIndex]
+	limitName := resource.Metadata[limitNameIndex]
+
+	limit := resource.Metadata[limitIndex]
+	usage := resource.Metadata[usageIndex]
```

Comment:
This looks like it could panic. Would be good to be a bit more defensive and have some error handling just in case.

Reply:
Fixed